### PR TITLE
Migrate to Firestore transactions for writes

### PR DESCRIPTION
**Pull request by Ehesp**
This PR introduces transaction based writes/deletes for extensions using Firestore, as discussed. 

Related PR: https://github.com/firebase/extensions/issues/48

### CODE REVIEW

File: delete-user-data/functions/src/index.ts [hunk 1]
```diff
@@ -122,10 +122,9 @@ const clearFirestoreData = async (firestorePaths: string, uid: string) => {
 
       if (!isRecursive) {
         logs.firestorePathDeleting(path, false);
-        await admin
-          .firestore()
-          .doc(path)
-          .delete();
+        await admin.firestore().runTransaction((async transaction => {
```
Comment (schmidt-sebastian) [inline delete-user-data/functions/src/index.ts]: Here and everywhere: Do you mind adding a comment on why you are doing this ("Wrapping in transaction to allow for automatic retries").

File: delete-user-data/functions/src/index.ts [hunk 2]
```diff
@@ -122,10 +122,9 @@ const clearFirestoreData = async (firestorePaths: string, uid: string) => {
...
+          return transaction.delete(admin.firestore().doc(path));
```
Comment (schmidt-sebastian) [inline delete-user-data/functions/src/index.ts]: Nit: Store `admin.firestore()` in a local variable.

File: firestore-send-email/functions/src/index.ts [hunk 3]
```diff
@@ -86,14 +86,16 @@ function validateFieldArray(field: string, array?: string[]) {
...
 async function processCreate(snap: FirebaseFirestore.DocumentSnapshot) {
-  return snap.ref.update({
-    delivery: {
-      startTime: admin.firestore.FieldValue.serverTimestamp(),
-      state: "PENDING",
-      attempts: 0,
-      error: null,
-    },
-  });
+  return admin.firestore().runTransaction((async transaction => {
```
Comment (schmidt-sebastian) [inline firestore-send-email/functions/src/index.ts]: Note sure if you have a specific style that you are following, but in general it seems like your functions don't need to be marked async.

File: delete-user-data/functions/src/index.ts (hunk 1 above)
Comment (Ehesp) [reply delete-user-data/functions/src/index.ts]: Sure thing, will get them added.

File: firestore-send-email/functions/src/index.ts (hunk 3 above)
Comment (Ehesp) [reply firestore-send-email/functions/src/index.ts]: This was to ensure the transaction returns a Promise. I'll change them to resolve a Promise from the callback instead.

**Comment by Ehesp** [comment]
@schmidt-sebastian I've updated the code with the feedback. I'll go through and re-test these are all working with the latest CLI release too. We'll also need a version bump before merging.

**Comment by laurenzlong** [comment]
I'll update the version numbers in a separate PR

**Pull request merged**